### PR TITLE
Repair quantized model export on a clean installation

### PR DESCRIPTION
**Pull request by damenknight**
## What changed

Two defects, either of which stops `waldo model export --quant` from running on
a working installation. The first gates the second: `ResolveRuntime` runs at
`internal/cli/model.go:939` and `calibration.Prepare` at `:962`, so the
calibration failure is unreachable until the probe is fixed. They are kept as
separate commits.

### 1. The quantizer is rejected for not implementing `--version`

`internal/modelquant/quant.go` resolved its tools by running `<tool> --version`
and rejecting anything that exited non-zero. `llama-quantize` parses its own
arguments and has no `--version`; it answers with usage text and status 1. The
other llama.cpp tools use the shared parser and do support it.

Measured against a local build of upstream `6f4f53f2b` carrying unrelated
local patches (they add flags and never touch version handling, so the sample
output below reports that build's own commit): `llama-cli`,
`llama-perplexity` and `llama-imatrix` each exit 0 and report a version, while
`llama-quantize` exits 1. That same binary then quantizes a 19,102,112 byte
model to 6,567,328 bytes and exits 0, so it works, it just cannot name itself.
The error told the operator to install llama.cpp and put `llama-quantize` on
PATH when it was already both.

This is not stale. On current master (`4df29be4f`),
`tools/quantize/quantize.cpp` still parses its own argv with a `strcmp` loop
whose flag list has no `--version`; the `llama_print_build_info` call it does
make runs during a quantize, not as a flag handler.

The probe moves into `probeVersion`, where every outcome is a named case: a
cancelled caller fails; a tool that answers is recorded; a tool that declines
the flag or does not answer within ten seconds is recorded without a version;
only a tool that could not be run at all is an error. The resolver already
computes a SHA-256 and the release BOM already carries it, so a tool that
reports no version stays identified by its digest:

```json
"quantizer": {"name": "llama-quantize", "version": "", "sha256": "0bef37e6..."}
```

`json:"version"` has no `omitempty`, so the key stays present and a consumer can
tell "no version reported" from "field absent".

Bounding the probe fixed a second thing. Resolution set no `WaitDelay`, so
`CombinedOutput` waited on the pipe rather than the process, and any background
child of the probe stalled the export for as long as it lived. With a wrapper
that leaves a 30-second child, resolution took 30.22s before and 2.02s after.

### 2. `--calibration` cannot create its own scratch directory

On a host where the scratch root does not exist yet, one flag apart:

```
--quant 4 --calibration core/common-pile/public-domain-review
  -> Error: stat /tmp/waldo-0/scratch: no such file or directory
--quant 4
  -> exported successfully
```

`os.MkdirTemp` requires its parent to exist, and the scratch root defaults
beneath the system temporary directory, so nothing guarantees it is there.

Two call sites already created that root, at two different modes: the lookaside
cache at 0700 before its download temporaries, and `model train-worker`, which
reached it as a side effect of `MkdirAll` on a deeper path at 0755. The mode of
a shared directory was decided by whichever command a host ran first.
`Cache.EnsureScratch` gives it one owner for both its existence and its mode,
and all three call sites use it.

The per-node and per-stage training directories beneath it tighten from 0755 to
0700 to match. Their consumers are same-user children: the training package
spawns workers with `exec.CommandContext` and `SysProcAttr{Setpgid: true}` and
nothing else, with no `Credential`, no setuid and no container, and both
directories are removed when the run ends.

### Why CI did not catch either

`resolveTool` had no coverage. Existing tests build `Tool` literals with the
unexported `path` field set directly, so they never resolve a real binary. The
only end-to-end test covering a quantized export is gated behind both
`command -v llama-quantize` and a Metal-capable MLX Python runtime
(`testing/e2e/model-mlx.sh:224`), so it skips on Linux CI and on any macOS
runner without MLX. Calibration is gated the same way.

## Verification

Byte counts, re-measured against the final branch:

| quant | bytes |
| --- | --- |
| source | 19,102,112 |
| Q8_0 | 10,161,568 |
| Q6_K | 9,590,176 |
| Q4_K_M | 6,567,328 |

Q4_K_M matches byte for byte what `llama-quantize` produces when driven by hand.

Calibrated export end to end on a Ryzen AI Max+ 395 (gfx1151), from a scratch
root that did not exist:

```
calibration   resolved core/common-pile/public-domain-review: 1 shards, 1.5M reference tokens
calibration   selected 100.0K byte tokens from 21 records in 1 shards
quantization  writing high-precision GGUF / measuring importance / quantizing as Q4_K_M
exported gguf model fix10m

quantizer  version=''                          sha256=0bef37e6...  (64 chars)
calibrator version='version: 9843 (4b72bb9d7)' sha256=6ae9286b...  (64 chars)
```

The quantizer reports nothing and stays pinned by digest. The calibrator is
recorded from the line that names a version, which matters because llama.cpp
answers on standard error and device diagnostics can appear ahead of it; that is
how a diagnostic line came to be recorded as the version during development. The
scratch root is created 0700 and left empty.

Ten tests cover the probe and its two helpers. Six fail against main's probe
semantics, including the reported bug, the diagnostic-line capture, and one that
waits the full 60 seconds on main for a tool that never answers. A seventh
passes on main but takes 30.23s there against 2.22s here, which is the pipe
defect above. `RejectsUnrunnableBinary` is a guard and passes either way. The
last two cover `toolVersion` and `bound`, helpers this PR introduces, and each
fails with its own fix reverted.

CI covers `ubuntu-latest` and `macos-latest`, so its four steps were run on both:
`gofmt`, `go vet ./...`, `go test ./...` across 23 packages, and
`ingest-direct.sh`. All pass, `-race` included. `./testing/all.sh` passes on
both; on Linux that also runs the real PyTorch lifecycle, which skips on macOS.
The four timing-sensitive tests ran 13 iterations and 52 executions, including
with every core saturated, without flaking or drifting.

`model train-worker` has no automated coverage, since
`model-torchtitan-multinode.sh` needs 2+ visible CUDA GPUs plus a TorchTitan
runtime and is opt-in behind `WALDO_E2E_MULTINODE=1`. Its scratch modes were
checked by running it against an unreachable rendezvous and inspecting the tree
while it polled: 0700 at every level including `node-1`, with `node-1` removed
on exit and the shared root left at 0700.

## Contracts

No durable format change and no ADR needed. `docs/COMPATIBILITY.md` requires an
ADR for a change to a persistent format's `kind` or `schema`, and neither moves.
`quantization.quantizer.version` keeps its type and its place in `BOM.json`, and
with no `omitempty` the key is present either way. Only the value changes: it
can now be empty, where before the export failed outright rather than producing
a BOM at all.

ADR 0027 says release evidence "hashes and identifies both executables". Both
still hold, since the SHA-256 and the name are unchanged, and it does not
mention a version field.

`Tool.Version` is documented as best effort in
`docs/TRAINING-AND-CALIBRATION.md`, since an empty version would otherwise read
as "this tool has no `--version`" when the host may just have been slow.
`internal/modelquant` is an internal package, which `docs/COMPATIBILITY.md`
lists as not a compatibility surface.

One behaviour change sits outside this PR's subject and is worth naming: the
per-node and per-stage `model train-worker` scratch directories tighten from
0755 to 0700. Commit 2 explains why.

## DCO

- [x] Every commit includes a `Signed-off-by` trailer (`git commit --signoff`).

### CODE REVIEW

File: internal/modelquant/quant.go [hunk 1]
```diff
@@ -138,17 +149,92 @@ func resolveTool(ctx context.Context, name string) (Tool, error) {
...
+	case probeCtx.Err() != nil:
+		// The probe outlived its own budget. A tool slow to answer must not
+		// fail an export, so it is recorded the same as one that cannot answer.
+		return "", nil
+	case errors.As(probeErr, &exitErr):
+		// llama-quantize parses its own arguments and does not implement
+		// --version, unlike the llama.cpp tools built on the shared parser, so
+		// it answers with usage text and a non-zero status. It reports no
+		// version; the digest still identifies it.
+		return "", nil
```
Comment (gmkurtzer) [inline internal/modelquant/quant.go]: **[P2] Do not accept every failed version probe.** Every `exec.ExitError` is treated as an unsupported `--version` flag. This also accepts crashes, signals, missing shared-library failures, and a broken `llama-imatrix`, postponing failure until after expensive export/calibration work. Limit the exception to the expected `llama-quantize` usage response; preserve errors for unexpected exit codes, signals, and tools that should support `--version`.

File: internal/lookaside/cache.go [hunk 2]
```diff
@@ -112,8 +112,12 @@ func DefaultCache() (*Cache, error) {
 
 func (cache *Cache) Root() string    { return cache.root }
 func (cache *Cache) Scratch() string { return cache.scratch }
-func (cache *Cache) MaxBytes() int64 { return cache.maxBytes }
-func (cache *Cache) Retained() bool  { return cache.retain }
+
+// EnsureScratch creates the scratch root. Callers that place temporary files
+// there own neither its location nor its mode.
+func (cache *Cache) EnsureScratch() error { return os.MkdirAll(cache.scratch, 0o700) }
```
Comment (gmkurtzer) [inline internal/lookaside/cache.go]: **[P2] Enforce permissions on an existing scratch root.** `MkdirAll(..., 0700)` applies the mode only when creating the directory. A scratch root created by an older command as 0755 remains 0755, despite `EnsureScratch` now owning its mode. After `MkdirAll`, chmod the scratch root to 0700 and add a test starting with an existing 0755 directory.

**Review by gmkurtzer — CHANGES_REQUESTED** [review]
I found two edge cases worth addressing before merge. The implementation otherwise looks solid, and the full Go suite, vet, repeated target-package tests, and target-package race tests pass.

**Comment by damenknight** [comment]
Both addressed, thanks. Also worth flagging: the first one caught a test of mine that was passing vacuously.

**Every failed probe accepted.** You're right that `ExitError` was doing far too much work. It excused crashes, signals, missing shared libraries, and a broken `llama-imatrix`, all deferred until after the export.

The exception is now limited to the tool that actually needs it, and to the status it actually returns:

```go
case declines && errors.As(probeErr, &exitErr) && exitErr.ExitCode() == toolUsageExit:
```

`declinesVersion` names `llama-quantize` alone, so any other tool refusing `--version` is an error rather than a quirk. `ExitError.ExitCode()` reports -1 for a signalled process, so a crash cannot reach that case without a separate signal check, and pinning the code to 1 rejects 127 and anything else unexpected. The reject path now carries the tool's own output, since "install llama.cpp" is unhelpful advice for a binary that is present but broken.

Two tests cover it: one asserts a crash, exit 127, exit 2, and even the bare usage status are all rejected for a tool expected to answer; the other asserts the usage refusal is excused for `llama-quantize` while a crash still is not.

This surfaced a real problem in `TestResolveToolBoundsReportedVersion`. It sets `PATH` to a temp directory, so its `awk` helper was never found, the fake tool exited 127, the old code silently excused it, and the assertion `len("") > 4096` passed for the wrong reason. It never exercised the bound at all. Rebuilt on shell builtins, it now feeds 262,144 bytes through and gets exactly 4,096 back.

I left the probe timeout returning no version rather than an error. A timeout is a statement about the host and not about the tool, and failing an export because a loaded machine was slow to answer `--version` seemed worse than recording the digest alone. Happy to tighten it if you disagree.

**Existing scratch root keeps its mode.** Correct, and `MkdirAll` silently no-ops on mode. `EnsureScratch` now chmods after creating, so a root left at 0755 by an older command is corrected rather than inherited. Test starts from an existing 0755 directory and asserts 0700 after.

Both commits amended rather than stacked, since the branch merges as-is. gofmt, vet, and `go test -race` clean, and CI's four steps run on Linux as well as macOS.

**Review by gmkurtzer — CHANGES_REQUESTED** [review]
The two findings from the earlier review are addressed:

- Version probing now accepts only the known llama-quantize usage refusal (exit 1), while crashes and other exit codes remain errors.
- EnsureScratch now tightens an existing scratch root to mode 0700, with regression coverage.

Validation passed: GitHub CI on macOS and Ubuntu, go test ./..., go vet ./..., and targeted race tests for modelquant, calibration, and lookaside.

One change is still required before merge: this branch conflicts with current main in internal/cli/model.go. During the rebase, please combine this PRs EnsureScratch/0700 changes with mains newer purge-on-success behavior. Remove the deferred PurgeUsed call and use this flow after run is defined:

    if err := runSecondaryStages(commandContext, cluster, modelRoot, scratch, cache, planWait, run, stdout, stderr); err != nil {
        return err
    }
    if _, err := cache.PurgeUsed(); err != nil {
        return fmt.Errorf("purge successful secondary training cache: %w", err)
    }
    return nil

This keeps failed training objects available for diagnosis and retry while purging after successful use.

**Comment by damenknight** [comment]
Rebased onto current main (`28eb8bc`) and resolved the `internal/cli/model.go` conflict as specified. The deferred `PurgeUsed` is gone and the flow now reads:

```go
if err := runSecondaryStages(commandContext, cluster, modelRoot, scratch, cache, planWait, run, stdout, stderr); err != nil {
    return err
}
if _, err := cache.PurgeUsed(); err != nil {
    return fmt.Errorf("purge successful secondary training cache: %w", err)
}
return nil
```

`EnsureScratch` and the 0700 per-node directory sit above it, using main's already-hoisted `DefaultCache`. I also confirmed the single-owner property still holds after your 58 commits: the four other `EffectiveScratchRoot` callers in `index_update.go` and `ingest.go` only pass it to `ValidateWorkLocations`, which creates nothing.

Two commits, still signed off, each building and testing on its own. `gofmt`, `go vet`, and `go test ./... -race` clean locally, and I ran CI's steps on Linux as well as macOS.

One note on the red checks: the failing step is `Local ingestion E2E`, and it is not from this branch. It fails identically on plain `origin/main` with no patches applied:

```
preflighting ingestion
Error: direct index ingest requires --title, --license, --source, --source-category, and --language
```

`main`'s own CI has been failing on this for at least the last five commits (`a36b5b22c`, `28eb8bc31`, `b3784d321`, `596f21cca`). On this PR the `gofmt`, `vet`, and `go test` steps all pass; only that E2E step fails. Happy to leave it or to look at it separately, whichever you prefer.

**Comment by gmkurtzer** [comment]
I went ahead and fixed and merged this. Thanks!
